pkg/roachpb/span_config.proto

```diff
@@ -0,0 +1,184 @@
+  // exchange, writes get pushed into the future and must wait on commit to
+  // ensure linearizability. For more, see #52745.
+  bool global_reads = 4;
+
+  // NumReplicas specifies the number of replicas, including both voting and
+  // non-voting kinds.
+  int32 num_replicas = 5;
+
+  // NumVoters specifies the number of voter replicas. If set to zero, we'll
+  // consider NumReplicas to be the voter replica count instead (i.e. no
+  // non-voting replicas).
+  int32 num_voters = 6;
+
+  // Constraints constrain which stores the both voting and non-voting replicas
+  // can be placed on.
+  //
+  // NB: The NumReplicas fields in Constraints must either:
+  // - add up to at most SpanConfig.NumReplicas (pinning the sum of replicas and
+  //   having the rest float freely);
+  // - all be zero (applying each constraint to all replicas).
+  repeated ConstraintsConjunction constraints = 7 [(gogoproto.nullable) = false];
+
+  // VoterConstraints constrains which stores the voting replicas can be placed
+  // on. This must be compatible with the Constraints field above, but not
+  // necessarily a subset. It's compatible as long as there are no prohibitive
+  // constraints above that are required here.
+  repeated ConstraintsConjunction voter_constraints = 8 [(gogoproto.nullable) = false];
+
+  // LeasePreference captures the preference for how range leases are to be
+  // placed. They're allowed to be placed elsewhere if needed, but will follow
+  // the stated preferences when possible.
+  //
+  // More than one lease preference is allowed; they're considered as the most
+  // preferred option to least. The first preference that an existing replica of
+  // a range matches will take priority for the lease.
+  repeated LeasePreference lease_preferences = 9 [(gogoproto.nullable) = false];
+}
+
+// SpanConfigEntry ties a span to its corresponding config.
+message SpanConfigEntry {
+  // Span is the keyspan the config is said to apply over.
+  Span span = 1 [(gogoproto.nullable) = false];
+
+  // Config is the set of attributes that apply over the corresponding keyspan.
+  SpanConfig config = 2 [(gogoproto.nullable) = false];
+};
+
+// GetSpanConfigsRequest is used to fetch the span configurations over the
+// specified keyspans.
+message GetSpanConfigsRequest {
+  // Spans to request the configurations for.
+  repeated Span spans = 1 [(gogoproto.nullable) = false];
+};
+
+// GetSpanConfigsResponse lists out the span configurations that overlap with
+// the requested spans.
+message GetSpanConfigsResponse {
+  message Result {
+    repeated SpanConfigEntry span_configs = 1 [(gogoproto.nullable) = false];
+  }
+
+  // Results capture the span configurations over the specified spans. There's
+  // one result for each Span in the matching GetSpanConfigsRequest, in the same
+  // order. A result will be present for each Span even if there are no configs
+  // to return.
+  repeated Result results = 2 [(gogoproto.nullable) = false];
```

Comment:
out of curiosity, is the client happy about this nesting? Would the client be happier with a flattened result set and over which it's likely going to be building a data structure anyway? It feels like there's a good bit of complexity added to deal with the nesting and unnesting when maybe everybody just wants it flattened.

Reply:
Hm yea the client doesn't really do much with the nesting; retrieved results are blindly plopped into another data structure. The other thing worth noting is that the spans requested aren't allowed to overlap anyway -- if the client wants a nested form, it could very well do that itself. Updating.

Reply:
Done.